dalm/training/generator_only/trainer.py

```diff
@@ -0,0 +1,274 @@
+
+from trl import SFTTrainer
+from trl.trainer import ConstantLengthDataset
+
+import argparse
+
+
+def create_datasets(
+    dataset_name: str,
+    split: str,
+    size_valid_set: Optional[int],
+    streaming: bool,
+    shuffle_buffer: int,
+    seq_length: int,
+    num_workers: int,
+    tokenizer: AutoTokenizer,
+    formatting_func: callable,
+):
+    dataset = load_dataset(
+        dataset_name,
+        split=split,
+        num_proc=num_workers if not streaming else None,
+        streaming=streaming,
+    )
+    if streaming:
+        print("Loading the dataset in streaming mode")
+        valid_data = dataset.take(size_valid_set)
+        train_data = dataset.skip(size_valid_set)
+        train_data = train_data.shuffle(buffer_size=shuffle_buffer, seed=None)
+    else:
+        dataset = dataset.train_test_split(test_size=0.05, seed=None)
+        train_data = dataset["train"]
+        valid_data = dataset["test"]
+        print(f"Size of the train set: {len(train_data)}. Size of the validation set: {len(valid_data)}")
+
+    chars_per_token = chars_token_ratio(train_data, tokenizer, formatting_func)
+    print(f"The character to token ratio of the dataset is: {chars_per_token:.2f}")
+
+    train_dataset = ConstantLengthDataset(
+        tokenizer,
+        train_data,
+        formatting_func=formatting_func,
+        infinite=True,
+        seq_length=seq_length,
+        chars_per_token=chars_per_token,
+    )
+    valid_dataset = ConstantLengthDataset(
+        tokenizer,
+        valid_data,
+        formatting_func=formatting_func,
+        infinite=False,
+        seq_length=seq_length,
+        chars_per_token=chars_per_token,
+    )
+    return train_dataset, valid_dataset
+
+
+def chars_token_ratio(dataset, tokenizer, formatting_func, nb_examples=400):
+    """
+    Estimate the average number of characters per token in the dataset.
+    """
+    total_characters, total_tokens = 0, 0
+    for _, example in tqdm(zip(range(nb_examples), iter(dataset)), total=nb_examples):
+        text = formatting_func(example)
+        total_characters += len(text)
+        if tokenizer.is_fast:
+            total_tokens += len(tokenizer(text).tokens())
+        else:
+            total_tokens += len(tokenizer.tokenize(text))
+
+    return total_characters / total_tokens
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--model_name", type=str, default="HuggingFaceH4/zephyr-7b-alpha", help="the model name")
+    parser.add_argument("--log_with", type=str, default="wandb", help="use 'wandb' to log with wandb")
+    parser.add_argument(
+        "--dataset_name", type=str, default="arcee-ai/azure-reading-comprehension-dataset", help="the dataset name"
+    )
+    parser.add_argument("--split", type=str, default="train", help="the split to use")
+    parser.add_argument("--size_valid_set", type=int, default=4000, help="the size of the validation set")
```

Comment:
It seems easier for users to express this in terms of percentage of training set rather than absolute size.  Eg, 20%

Reply:
I believe this argument for streaming, I believe in this scenario a preset size is required

Reply:
Ah got it, for streaming we don't know the full size of the dataset until we load it.  I think its confusing and misleading though to advertise this parameter without mentioning the fact that its streaming only.  

WDYT of changing to this?

```
    parser.add_argument(
      "--size_valid_set_streaming", 
      type=int, 
      default=4000, 
      help="the size of the validation set when used in streaming mode, ignored otherwise"
    )

```